src/marqo/tensor_search/tensor_search.py

```diff
@@ -1235,6 +1346,257 @@ def _lexical_search(
+    qidx_to_vectors: Dict[Qidx, List[List[float]]] = defaultdict(list)
+    for qidx, ptr in qidx_to_job.items():
+        vectors = job_ptr_to_vectors[ptr.job_hash][ptr.start_idx: ptr.end_idx]
+        qidx_to_vectors[qidx].append(vectors)
+
+        q = queries[qidx]
+        ordered_queries = list(q.q.items()) if isinstance(q.q, dict) else None
+        if ordered_queries:
+            # multiple queries. We have to weight and combine them:
+            index_info = get_index_info(config=config, index_name=q.index)
+            weighted_vectors = [np.asarray(vec) * weight for vec, weight in zip(vectors, [w for _, w in ordered_queries])]
+            vectorised_text = np.mean(weighted_vectors, axis=0)
+            if index_info.index_settings['index_defaults']['normalize_embeddings']:
+                norm = np.linalg.norm(vectorised_text, axis=-1, keepdims=True)
+                if norm > 0:
+                    vectorised_text /= np.linalg.norm(vectorised_text, axis=-1, keepdims=True)
+            qidx_to_vectors[qidx] = list(vectorised_text)
+        else:
+            qidx_to_vectors[qidx] = vectors[0]
+
+    ## 4. Create msearch request bodies and combine to aggregate.
+    query_to_body_parts: Dict[Qidx, List[Dict]] = {}
+    query_to_body_count: Dict[Qidx, int] = {} # Keep track of count, so we can separate after msearch call.
+    for qidx, qq in enumerate(queries):
+        index_info = get_index_info(config=config, index_name=q.index)
+        if q.filter is not None:
+            contextualised_filter = utils.contextualise_filter(
+                filter_string=q.filter,
+                simple_properties=index_info.get_text_properties())
+        else:
+            contextualised_filter = ''
+
+        vector_properties_to_search = get_vector_properties_to_search(qq.searchableAttributes, index_info, raise_on_searchable_attribs)
+        vectorised_text = qidx_to_vectors[qidx]
+        body = construct_msearch_body_elements(vector_properties_to_search, qq.offset, q.filter, index_info, qq.limit, vectorised_text, qq.attributesToRetrieve, qq.index, contextualised_filter)
+        query_to_body_parts[qidx] = body
+        query_to_body_count[qidx] = len(body)
+
+    aggregate_body = [line for body in query_to_body_parts.values() for line in body]
+    if not aggregate_body:
+        return []
+
+    logger.info(f"search (tensor) pre-processing: took {(timer() - start_preprocessing_time):.3f}s to vectorize and process query.")
+
+    ## 5. POST aggregate  to /_msearch
+    responses = bulk_msearch(config, aggregate_body)
+    start_postprocess_time = timer()
+
+    # 6. Get documents back to each query, perform "gather" operation
+    results = []
+    for qidx, count in query_to_body_count.items():
+        num_of_docs = count // 2
+        result = responses[:num_of_docs]
+        responses = responses[num_of_docs:] # remove docs from response for next query
```

Comment:
mutating the responses in this loop makes me a bit uncomfortable

Reply:
It made me too, I made it a deep copy now.